tensorflow/python/framework/python_op_gen.cc

```diff
@@ -367,6 +405,155 @@ string GenEagerPythonOp::Code() {
+      strings::StrAppend(&type_annotation, "_ops.Tensor[", param_type_map[arg.type_attr()], "]");
+    } else {
+      // Get the dtype of the tensor
+      const string py_dtype = python_op_gen_internal::DataTypeToPython(arg.type(), "_dtypes.");
+      if (dtypes_map.find(py_dtype) != dtypes_map.end()) {
+        strings::StrAppend(&type_annotation, "_ops.Tensor[", dtypes_map[py_dtype], "]");
+      }
+    }
+
+    param_type_map[arg.name()] = type_annotation;
+  }
+
+  // Add type annotations to params
+  string parameters;
+  for (const auto& param : params_no_default_) {
+    if (!parameters.empty()) strings::StrAppend(&parameters, ", ");
+    strings::StrAppend(&parameters, param.GetRenameTo());
+
+    if (param_type_map.find(param.GetName()) != param_type_map.end()) {
+      if(!param_type_map[param.GetName()].empty()){
+        strings::StrAppend(&parameters, ": ", param_type_map[param.GetName()]);
+      }
+    }
+  }
+
+  return parameters;
+}
+
+// Generate TypeVars using attrs
+void GenEagerPythonOp::GenerateTypeVars() {
+  bool added_typevar = false;
+  for (int i = 0; i<op_def_.attr_size(); ++i) {
+    const auto& attr(op_def_.attr(i));
+    if (attr.type() == "type") {
+      std::vector<string> allowed_types;
+      bool has_dtype_half = false;
+      for (int t : attr.allowed_values().list().type()) {
+        if (t == 19) { // DT_HALF = 19
+          has_dtype_half = true;
+          break;
+        }
+        DataType dtype = static_cast<DataType>(t);
+        const string py_dtype = python_op_gen_internal::DataTypeToPython(dtype, "_dtypes.");
+        if (dtypes_map.find(py_dtype) != dtypes_map.end()) {
+          allowed_types.emplace_back(dtypes_map[py_dtype]);
+        }
+      }
+
+      // Do not create a type variable that includes the dtype half
+      if (has_dtype_half) continue;
+
+      // If all dtypes are allowed, add them all
+      if (allowed_types.empty()) {
+        for (std::pair<string, string> map_dtype : dtypes_map) {
+          allowed_types.emplace_back(map_dtype.second);
+        }
+      }
+
+      std::sort(allowed_types.begin(), allowed_types.end());
+
+      string typevar_dtypes;
+      for (std::vector<string>::iterator it = allowed_types.begin(); it != allowed_types.end(); ++it) {
+        if (!typevar_dtypes.empty()) strings::StrAppend(&typevar_dtypes, ", ");
+        strings::StrAppend(&typevar_dtypes, *it);
+      }
+
+      const string type_var_name = "TV_" + op_def_.name() + "_" + attr.name();
+      strings::StrAppend(&result_, type_var_name, " = TypeVar(\"", type_var_name, "\", ", typevar_dtypes,")\n");
+      added_typevar = true;
+    }
+  }
+
+  if(added_typevar) strings::StrAppend(&result_, "\n");
+}
+
+void GenEagerPythonOp::AddReturnTypeAnnotation() {
+  string return_type = "";
+  if (op_def_.output_arg_size() == 1) {
+    const auto& arg = op_def_.output_arg(0);
+    // If the "type" field is set, the return Tensor has a single DataType
+    if (arg.type() != 0) {
```

Comment:
There seems to be some redundancy here as well. Could we instead have a function that extracts the annotation code? Something like:

strings::StrAppend(&return_type, TypeAnnotationFor(arg));

Reply:
Changed this to use a map to get the type annotation which cleaned up this function a lot.